src/python/pants/engine/rules.py

```diff
@@ -405,6 +405,10 @@ def __new__(cls,
               dependency_optionables=None,
               cacheable=True):
 
+    for selector in input_selectors:
+      if not isinstance(selector, type):
+        raise ValueError('Parameter selectors must be types: got {}'.format(selector))
+
```

Comment:
You can actually do this as of #7115 :
```python
from pants.util.objects import Exactly, TypedCollection, datatype

class TaskRule(datatype([
  ('input_selectors', TypedCollection(Exactly(type))),
  # ...
```

The error message should include the field which failed to type check, so this should be an equivalent edit.